### PR TITLE
feat/ balancer connector

### PR DESCRIPTION
**Pull request by vic-en**
**Before submitting this PR, please make sure**:

- [x] Your code builds clean without any errors or warnings
- [x] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:
Balancer connector.

**Tests performed by the developer**:
- Curl tests
- Unit test

**Tips for QA testing**:
- Curl test
- End-to-end testing with Hummingbot client.

### CODE REVIEW

**Comment by nikspz** [comment]
@vic-en Could you please fix failing tests/ add coverage?
https://github.com/hummingbot/gateway/actions/runs/7889551246/job/21530292736?pr=280

**Comment by nikspz** [comment]
- Latest development branch + https://github.com/hummingbot/gateway/pull/280
    - Test performed:
        - gateway generate certs and yarn start = ok
            - ethereum arbitrum
                - gateway connect balancer = ok
                - checked balance successfully
                - approved WETH and USDC using Client interface
                - create/started amm_arb using balancer ethereum arbitrum
                    - checked status - got warning: Markets are offline for WETH-USDC
                        - Got not realistic PNL for balancer_ethereum_arbitrum to paper_trade
                        - amm/price failed during checking status
                    - curl
                        - amm/price: ok
                        - Reviewed delay before response is about ~10-15 sec
            - ethereum mainnet
                - gateway connect balancer = ok
                - checked balance successfully
                - created amm_arb using balancer ethereum mainnet
            - avalanche_avalanche
                - gateway connect balancer = ok
                - checked balance successfully
                - created amm_arb using balancer avalanche_avalanche
            - polygon_mainnet
                - gateway connect balancer = ok
                - checked balance successfully
                - created amm_arb using balancer polygon_mainnet

image

Was able to get 1 trade on balancer arbitrum
balancer
amm/price failed during checking status
image

Reviewed delay before response is about ~10-15 sec
balancer amm price

**Comment by nikspz** [comment]
@vic-en Could you please fix failing tests/ add coverage for us to be able to check and add it to this release?
https://github.com/hummingbot/gateway/actions/runs/7889551246/job/21530292736?pr=280

**Comment by vic-en** [comment]
@nikspz I'm re-running the tests. I don't think the faillure is due to my change.

**Comment by vic-en** [comment]
FYI @fengtality @nikspz 
I believe the workflow file was updated at some point. 
I see only ethereum goerli node url is now patched - https://github.com/hummingbot/gateway/blob/be368edc8cfde1d8ebd36bd4aac0bb52a644f593/.github/workflows/workflow.yml#L81

That implies several unit tests will be making call to real node endpoints as defined the the configuration templates.

Aside that, I see several unit tests referencing mainnet. Those are not ideal, especially when no patch in workflow file.

image

**Comment by nikspz** [comment]
We should try with router disabled but we need a the config for that
image
we have this thing for Uniswap
image

@vic-en  Did you have any chance to check this feature for balancer? Could you please provide some update?

**Comment by fengtality** [comment]
@vic-en I think the hit-and-miss nature of the router means that this connector will need to support the `useRouter: false` flag to be usable by the community. I think everyone does the same thing with the Uniswap connector.

**Comment by nikspz** [comment]
hi @vic-en please fix branch conflicts, update with latest development and continue with this PR

**Comment by vic-en** [comment]
@nikspz This is ready.

**Comment by nikspz** [comment]
@vic-en Could you please check failing tests / add coverage?

**Comment by fengtality** [comment]
I think this PR fixes the failing tests: https://github.com/hummingbot/gateway/pull/311

**Review by fengtality — APPROVED** [review]

**Comment by nikspz** [comment]
- Test performed:
    - commit 65f4d94
    - Cloned and installed balancer gw280 + latest development Client
    - gateway generate certs and review gateway ONLINE
    - connected balancer connector for
        - ethereum mainnet: ok
        - ethereum arbitrum: ok
        - avalanche: ok
        - polygon: ok
    - chain/status: ok
    - wallet/add: ok
    - amm/price:
        - ethereum mainnet: ok
            - able to get the price
            - some of requests returns `Error: Price query failed: No pool found`, however checking again after short time was successfull
        - ethereum arbitrum: ok
        - polygon: ok
        - avalanche: "Error: 503 unsupported chain or connector


@vic-en Could you check avalanche? Failed to amm/price for balancer_avalanche_avalanche

image

image

image
image

**Comment by vic-en** [comment]
@nikspz This is updated. Avalanche now works.

Also, to test with poolId just add the extra poolId in the query like below
image
Note that you can get poolId from url on that Balancer page 
image

cc @fengtality

**Comment by nikspz** [comment]
@vic-en 

avalanche_avalanche - trade was successful ✔
https://snowtrace.io/tx/0x53483b9fc56f1cf6d7d9c535295372ca04220743ad2f393b2685e97673248688?chainId=43114

amm/trade failing for ethereum_arbitrum ❌
Upon checking reason was - balancer ethereum_arbitrum failed to approve through Client ❌ (Though it returned approved before)
image

[logs_gateway_app.log.2024-04-22.log](https://github.com/hummingbot/gateway/files/15065525/logs_gateway_app.log.2024-04-22.log)
image
https://arbiscan.io/tx/0xd5e5a6c588793f019b58d946e3352092c6d6c906ae9b0fd3eb81f3e664b32835
image

liquidity/add ❌
"TypeError: uniswapish.addPosition is not a function\n
image
image

**Comment by vic-en** [comment]
@nikspz The LP endpoints for uniswap like contracts cannot work with Balancer. Balancer is different.
The poolId I added is to make swaps to interract with only the pool whose Id is supplied.

Note:
- I'll look at the issue with arbitrum approval.
- I'll also create a pr on the client to make amm arb configure a poolId
Eta is on/before this weekend.

**Comment by nikspz** [comment]
> Note:
> 
> * I'll look at the issue with arbitrum approval.
> * I'll also create a pr on the client to make amm arb configure a poolId
>   Eta is on/before this weekend.

Hi @vic-en Do you have any updates on this one?

**Comment by vic-en** [comment]
@nikspz I'll push the update today.

**Comment by vic-en** [comment]
@nikspz This is ready for testing.
Regarding the failed swaps on arbitrum, they failed due to slippage. Increase the slippage in the configuration file.

File: src/connectors/balancer/balancer.ts [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+        {
+            tokenIn: baseToken.address,
+            tokenOut: quoteToken.address,
+            amount: amount,
+            gasPrice: BigNumber.from(this._chain.gasPrice.toFixed(0)),
+        }
+      );
+      if (info.swaps.length === 0) {
+        throw new UniswapishPriceError(
+          `No pool found for ${quoteToken.address} to ${baseToken.address}.`
```
Comment (fengtality) [inline src/connectors/balancer/balancer.ts:240]: this should be `No pool found for ${baseToken.address} to ${quoteToken.address}.`

**Review by fengtality — APPROVED** [review]
Approved - Note small typo in logging statement @vic-en

**Comment by fengtality** [comment]
I will merge this PR into `development` and address the typo there. The failing tests are due to the Uniswap router used in the tests requiring network access and can be ignored for now.

**Pull request merged**